src/protocol/serialize_stream.rs

```diff
@@ -1,426 +1,324 @@
-    {
-        let mut ws = SerializeStream::new(&mut page, 0);
-        ws.write_int(42).expect("Should not fail");
-    }
-
-    let mut rs = DeserializeStream::new(&mut page, 0);
-    let val = rs.read_int().unwrap();
-
-    assert_eq!(val, 42);
-}
-
-#[test]
-fn write_read_several_ints() {
-    let mut page = MemoryPage::new(4 + 4 + 4 + 3);
-    {
-        let mut ws = SerializeStream::new(&mut page, 0);
-
-        ws.write_int(2093608745).expect("Should not fail");
-        ws.write_int(0).expect("Should not fail");
-        ws.write_int(-1294).expect("Should not fail");
-    }
-
-    let mut rs = DeserializeStream::new(&mut page, 0);
-    assert_eq!(rs.read_int().unwrap(), 2093608745);
-    assert_eq!(rs.read_int().unwrap(), 0);
-    assert_eq!(rs.read_int().unwrap(), -1294);
-}
+use data_type::DataType;
+use storage::MemoryPage;
+use protocol::packer;
+
+/// Serializes and writes values to MemoryPage.
+/// Always writes values in the architertural endian currently.
+#[derive(Debug)]
+pub struct SerializeStream<'a> {
+    position: usize,
+    page: &'a mut MemoryPage,
+}
+
+impl<'a> SerializeStream<'a> {
+    /// Creates new stream.
+    pub fn new(page: &'a mut MemoryPage, pos: usize) -> Self {
+        SerializeStream {
+            position: pos,
+            page: page,
+        }
+    }
+
+    fn check_available_space(&self, val: isize) -> Result<(), String> {
+        let avail: isize = self.page.data().len() as isize - self.position as isize;
+
+        if avail < val as isize {
+            Err(format!("Not enough memory to serialize value: required={}, available={}",
+                        val,
+                        avail))
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Check if we have enough memory in page to write a static typed value.
+    fn check_static_type_len(&self, field_type: DataType) -> Result<(), String> {
+        let field_len: isize = field_type.static_len() as isize;
+
+        match self.check_available_space(field_len) {
+            Err(e) => return Err(format!("Unable to write {:?}: {:?}", field_type, e)),
+            Ok(_) => ()
+        }
+
+        Ok(())
+    }
+
+    /// Check if we have enough memory in page to write a dynamic typed value.
+    fn check_dynamic_type_len(&self, field_type: DataType, size: usize) -> Result<(), String> {
+        let field_len: isize = field_type.static_len() as isize;
+
+        match self.check_available_space(field_len + size as isize) {
+            Err(e) => return Err(format!("Unable to write {:?}: {:?}", field_type, e)),
+            Ok(_) => ()
+        }
+
+        Ok(())
+    }
+
+    /// Write INTEGER to stream.
+    pub fn write_int(&mut self, val: i32) -> Result<(), String> {
+        self.check_static_type_len(DataType::INTEGER)?;
+
+        let mem = self.page.data_mut();
+        let len = mem.len();
+
+        packer::pack_int(&mut mem[self.position..len], val);
+
+        self.position += 4;

```

Comment:
`DataType::XXX::static_len()` here and in other places. That's why I said I don't like `check_len` functions - you still need to calculate sizes in parental function.